src/watersensor/waterSensor.cpp

```diff
@@ -0,0 +1,169 @@
+#include "waterSensor.hpp"
+#include "product.hpp" // Added by PJB. Is it conventional to do this? Not sure but we need USB_PWR_DETECT_PIN
+#include "cli/conio.hpp"
+
+WaterSensor::WaterSensor(uint8_t water_detect_en_pin_to_set, uint8_t water_detect_pin_to_set, uint8_t window_size, uint8_t array_size) : water_detect_en_pin(water_detect_en_pin_to_set), water_detect_pin(water_detect_pin_to_set), moving_window_size(window_size), water_detect_array_size(array_size)
+{
+    water_detect_array = (uint8_t *)calloc(water_detect_array_size, sizeof(uint8_t));
+}
+
+WaterSensor::~WaterSensor()
+{
+    free(water_detect_array);
+}
+
+// reset the array, load it all up w/ 0s and clear the sum & location
+bool WaterSensor::resetArray()
+{
+    // clear the buffer out
+    memset(water_detect_array, (uint8_t)0, (sizeof(uint8_t) * water_detect_array_size));
+    array_sum = 0;
+    array_location = 0;
+    samples_taken_since_reset = 0;
+    return 0;
+}
+
+void WaterSensor::setWindowSize(uint8_t window_size_to_set)
+{
+    // swtich the window parameter and clear the sum (for resumming)
+    moving_window_size = window_size_to_set;
+    array_sum = 0;
+
+    // sum all of the array items from the current location backward for the entire window,
+    //  or until you get the number of samples taken after reset
+    for (uint8_t i = 0; (i < moving_window_size) && (i < samples_taken_since_reset); i++)
+    {
+        array_sum += water_detect_array[waterDetectArrayLocation(array_location, (-1 * i))];
+    }
+    // if changing from a larger to smaller sample window size, change samples_taken_since_reset accordingly
+    if (samples_taken_since_reset > moving_window_size)
+        samples_taken_since_reset = moving_window_size;
+}
+
+uint8_t WaterSensor::takeReading()
+{
+    // increment array location
+    array_location = (array_location + 1) % water_detect_array_size;
+    // subtract last value in the window from the rolling sum
+    array_sum -= water_detect_array[waterDetectArrayLocation(array_location, (-1 * moving_window_size))];
+
+    digitalWrite(water_detect_en_pin, LOW);
+    delayMicroseconds(WATER_DETECT_EN_TIME_US);
+    water_detect_array[array_location] = digitalRead(water_detect_pin); // comment out if using above logic block
+    digitalWrite(water_detect_en_pin, HIGH);
```

Comment:
Note that current measurement on a PicoScope indicates a neg pulse width of 1.010 ms with 1.289 us standard deviation over 100 measurements.

Reply:
This should use `WaterSensor::getCurrentReading`